addons/mail/static/src/messaging/entity/thread/thread.js

```diff
@@ -3,11 +3,113 @@ odoo.define('mail.messaging.entity.Thread', function (require) {
 
 function ThreadFactory({ Entity }) {
 
     class Thread extends Entity {
 
+        /**
+         * @override
+         */
+        init(...args) {
+            super.init(...args);
+            /**
+             * Timer of current partner that was currently typing something, but
+             * there is no change on the input for several time. This is used
+             * in order to automatically notify other members that current
+             * partner has stopped typing something, due to making no changes
+             * on the composer for some time.
+             */
+            this._currentPartnerInactiveTypingTimer = new Timer(
+                this.env,
+                () => this.async(() => this._onCurrentPartnerInactiveTypingTimeout()),
+                5 * 1000
+            );
+            /**
+             * Last 'is_typing' status of current partner that has been notified
+             * to other members. Useful to prevent spamming typing notifications
+             * to other members if it hasn't changed. An exception is the
+             * current partner long typing scenario where current partner has
+             * to re-send the same typing notification from time to time, so
+             * that other members do not assume he/she is no longer typing
+             * something from not receiving any typing notifications for a
+             * very long time.
+             *
+             * Supported values: true/false/undefined.
+             * undefined makes only sense initially and during current partner
+             * long typing timeout flow.
+             */
+            this._currentPartnerLastNotifiedIsTyping = undefined;
+            /**
+             * Timer of current partner that is typing a very long text. When
+             * the other members do not receive any typing notification for a
+             * long time, they must assume that the related partner is no longer
+             * typing something (e.g. they have closed the browser tab).
+             * This is a timer to let other members know that current partner
+             * is still typing something, so that they should not assume he/she
+             * has stopped typing something.
+             */
+            this._currentPartnerLongTypingTimer = new Timer(
+                this.env,
+                () => this.async(() => this._onCurrentPartnerLongTypingTimeout()),
+                50 * 1000
+            );
+            /**
+             * Determines whether the next request to notify current partner
+             * typing status should always result to making RPC, regardless of
+             * whether last notified current partner typing status is the same.
+             * Most of the time we do not want to notify if value hasn't
+             * changed, exception being the long typing scenario of current
+             * partner.
+             */
+            this._forceNotifyNextCurrentPartnerTypingStatus = false;
```

Comment:
Why having this variable? Why do a force at next send flow, instead of a send immediately at long timeout?

Reply:
If throttle is in cooling down phase, then the long typing timeout has to wait that long. Other members could assume in the meantime that this partner is no longer typing, and create a small flickering.

It shouldn't really happen in practice with how timers are set though: 50 seconds long typing, 60 seconds other members inactive typing, 2.5 seconds throttle function. This flicker could theoretically happen if network + server + client-side processing of notification takes more than 7.5 seconds, quite unlikely to happen in practice.

Reply:
Ok, just change the reset call as discussed, and add the rest as comment in the code.